File: app/src/main/java/com/inspiredandroid/linuxcommandbibliotheca/misc/FragmentCoordinator.java
package com.inspiredandroid.linuxcommandbibliotheca.misc;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.view.View;
import android.view.ViewGroup;

import com.inspiredandroid.linuxcommandbibliotheca.CommandManActivity;
import com.inspiredandroid.linuxcommandbibliotheca.R;
import com.inspiredandroid.linuxcommandbibliotheca.ScriptChildrenActivity;
import com.inspiredandroid.linuxcommandbibliotheca.fragments.BasicChildrenFragment;
import com.inspiredandroid.linuxcommandbibliotheca.fragments.CommandManFragment;
import com.inspiredandroid.linuxcommandbibliotheca.models.Command;

import io.realm.Realm;

/**
 * Created by simon on 25/01/17.
 */
public class FragmentCoordinator {

    public static void startScriptCategoryActivity(FragmentActivity activity, int category) {
        if (!isTabletLayout(activity)) {
            Intent intent = new Intent(activity, ScriptChildrenActivity.class);
            intent.putExtra(ScriptChildrenActivity.Companion.getEXTRA_CATEGORY_ID(), category);
            activity.startActivity(intent);
        } else {
            View container = activity.findViewById(R.id.fragment_container_secondary);
            ((ViewGroup) container).removeAllViews();

            activity.getIntent().putExtra(ScriptChildrenActivity.Companion.getEXTRA_CATEGORY_ID(), category);

            Fragment fragment = new BasicChildrenFragment();

            FragmentManager fragmentManager = activity.getSupportFragmentManager();
            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

            fragmentTransaction.replace(R.id.fragment_container_secondary, fragment);
            fragmentTransaction.commitAllowingStateLoss();
        }
    }

    public static void startCommandManActivity(FragmentActivity activity, long id) {
        if (!isTabletLayout(activity)) {
            Intent intent = new Intent(activity, CommandManActivity.class);
            intent.putExtra(CommandManActivity.Companion.getEXTRA_COMMAND_ID(), id);
            activity.startActivity(intent);
        } else {
            View container = activity.findViewById(R.id.fragment_container_secondary);
            ((ViewGroup) container).removeAllViews();

            Realm realm = Realm.getDefaultInstance();

            Command command = realm.where(Command.class).equalTo(Command.Companion.getID(), id).findFirst();
            String name = command.getName().toUpperCase();
            int category = command.getCategory();

            realm.close();

            showManFragmentInSecondaryFrame(activity, name, id, category);
        }
    }

    public static void startCommandManActivity(FragmentActivity activity, String name) {
        Realm realm = Realm.getDefaultInstance();
        Command command = realm.where(Command.class).equalTo(Command.Companion.getNAME(), name).findFirst();
        if(command == null) {
            realm.close();
            return;
        }
        long id = command.getId();
        int category = command.getCategory();
        realm.close();

        if (!isTabletLayout(activity)) {
            Intent intent = new Intent(activity, CommandManActivity.class);
            intent.putExtra(CommandManActivity.Companion.getEXTRA_COMMAND_ID(), id);
            activity.startActivity(intent);
        } else {
            View container = activity.findViewById(R.id.fragment_container_secondary);
            ((ViewGroup) container).removeAllViews();

            showManFragmentInSecondaryFrame(activity, name, id, category);
        }
    }

    /**
     * @param name     command name
     * @param id       unique command id
     * @param category command category
     */
    private static void showManFragmentInSecondaryFrame(FragmentActivity activity, String name, long id, int category) {
        Fragment fragment = new CommandManFragment();

        // Add unique command ID for fragment
        Bundle bundle = new Bundle();
        bundle.putLong(CommandManActivity.Companion.getEXTRA_COMMAND_ID(), id);
        bundle.putString(CommandManActivity.Companion.getEXTRA_COMMAND_NAME(), name);
        bundle.putInt(CommandManActivity.Companion.getEXTRA_COMMAND_CATEGORY(), category);
        fragment.setArguments(bundle);

        FragmentManager fragmentManager = activity.getSupportFragmentManager();
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();

        fragmentTransaction.replace(R.id.fragment_container_secondary, fragment);
        fragmentTransaction.commitAllowingStateLoss();
    }

    public static boolean isTabletLayout(Activity activity) {
        return activity.findViewById(R.id.fragment_container_secondary) != null;
    }
}

File: app/src/main/java/com/inspiredandroid/linuxcommandbibliotheca/misc/FragmentCoordinator.kt
package com.inspiredandroid.linuxcommandbibliotheca.misc

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import com.inspiredandroid.linuxcommandbibliotheca.R
import com.inspiredandroid.linuxcommandbibliotheca.activities.BasicGroupActivity
import com.inspiredandroid.linuxcommandbibliotheca.activities.CommandManActivity
import com.inspiredandroid.linuxcommandbibliotheca.fragments.BasicGroupFragment
import com.inspiredandroid.linuxcommandbibliotheca.fragments.CommandManFragment
import com.inspiredandroid.linuxcommandbibliotheca.models.Command
import io.realm.Realm
import io.realm.kotlin.where

/* Copyright 2019 Simon Schubert
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/
object FragmentCoordinator {

    fun startScriptCategoryActivity(activity: FragmentActivity?, category: Int) {
        activity?.let {
            if (!isTabletLayout(activity)) {
                val intent = Intent(activity, BasicGroupActivity::class.java)
                intent.putExtra(BasicGroupActivity.EXTRA_CATEGORY_ID, category)
                activity.startActivity(intent)
            } else {
                val container = activity.findViewById<View>(R.id.fragment_container_secondary)
                (container as ViewGroup).removeAllViews()

                activity.intent.putExtra(BasicGroupActivity.EXTRA_CATEGORY_ID, category)

                val fragment = BasicGroupFragment()

                val fragmentManager = activity.supportFragmentManager
                val fragmentTransaction = fragmentManager.beginTransaction()

                fragmentTransaction.replace(R.id.fragment_container_secondary, fragment)
                fragmentTransaction.commitAllowingStateLoss()
            }
        }
    }

    fun startCommandManActivity(activity: FragmentActivity?, id: Long) {
        activity?.let {
            if (!isTabletLayout(activity)) {
                val intent = Intent(activity, CommandManActivity::class.java)
                intent.putExtra(CommandManActivity.EXTRA_COMMAND_ID, id)
                activity.startActivity(intent)
            } else {
                val container = activity.findViewById<View>(R.id.fragment_container_secondary)
                (container as ViewGroup).removeAllViews()

                val realm = Realm.getDefaultInstance()

                val command = realm.where<Command>().equalTo(Command.ID, id).findFirst()
                if (command != null) {
                    showManFragmentInSecondaryFrame(activity, id)
                }

                realm.close()
            }
        }
    }

    fun startCommandManActivity(activity: FragmentActivity?, name: String) {
        activity?.let {
            val realm = Realm.getDefaultInstance()
            val command = realm.where<Command>().equalTo(Command.NAME, name).findFirst()
            if (command == null) {
                realm.close()
                return
            }
            val id = command.id.toLong()
            realm.close()

            if (!isTabletLayout(activity)) {
                val intent = Intent(activity, CommandManActivity::class.java)
                intent.putExtra(CommandManActivity.EXTRA_COMMAND_ID, id)
                activity.startActivity(intent)
            } else {
                val container = activity.findViewById<View>(R.id.fragment_container_secondary)
                (container as ViewGroup).removeAllViews()

                showManFragmentInSecondaryFrame(activity, id)
            }
        }
    }

    private fun showManFragmentInSecondaryFrame(activity: FragmentActivity, id: Long) {
        val fragment = CommandManFragment()

        // Add unique command ID for fragment
        val bundle = Bundle()
        bundle.putLong(CommandManActivity.EXTRA_COMMAND_ID, id)
        fragment.arguments = bundle

        val fragmentManager = activity.supportFragmentManager
        val fragmentTransaction = fragmentManager.beginTransaction()

        fragmentTransaction.replace(R.id.fragment_container_secondary, fragment)
        fragmentTransaction.commitAllowingStateLoss()
    }

    fun isTabletLayout(activity: Activity?): Boolean {
        return if (activity == null) {
            false
        } else {
            activity.findViewById<View>(R.id.fragment_container_secondary) != null
        }
    }
}
